modules/concurrency/include/hephaestus/concurrency/when_all_range.h

```diff
@@ -0,0 +1,175 @@
+//=================================================================================================
+// Copyright (C) 2023-2025 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <memory>
+#include <optional>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+
+#include <stdexec/__detail/__execution_fwd.hpp>
+#include <stdexec/__detail/__senders_core.hpp>
+#include <stdexec/execution.hpp>
+
+namespace heph::concurrency {
+template <typename Range>
+concept SenderRange = std::ranges::range<Range> && stdexec::sender<std::ranges::range_value_t<Range>>;
+
+namespace internal {
+
+struct WhenAllStopCallback {
+  stdexec::inplace_stop_source* stop_source;
+  void operator()() const {
+    stop_source->request_stop();
+  }
+};
+
+enum WhenAllRangeState : std::uint8_t { STARTED, ERROR, STOPPED };
```

Comment:
This leaks the `STARTED`, `ERROR`, `STOPPED` names in the `heph::concurrency::internal` namespace. I am assuming a `enum class` was not used because the enumeration is used as an atomic type, correct?

If so, please use this scoped traditional enum pattern:

```cpp
struct WhenAllRangeState
{
    enum Enum : std::uint8_t { STARTED, ERROR, STOPPED };
};
```